Fixes #4435, Refactors Rejuvenation, Fixes Flipping Runtime

KasparoVy opened the pull request:
:cl:
fix: Rejuvenating a mob now unbuckles them correctly (if they're buckled to something).
fix: Rejuvenating a mob will no longer remove their hair.
fix: Rejuvenating a mob will no longer permanently blacken their eyes.
fix: Rejuvenating a mob will no longer remove their prostheses with fleshy limbs, it will just fix the prostheses.
fix: Rejuvenating a mob will not remove organs/limbs belonging to other species if said organs/limbs are in the same place(s) as the mob's default organs/limbs would be.
fix: Fixes a runtime when trying to flip with an object other than a grab in your hand.
fix: Doing a 'cybernetic repair' procedure on the head of a disfigured IPC will re-figure them.
tweak: Rejuvenating a mob will only regrow organs/limbs that are missing (by comparison to the standard organs for their species).
tweak: The regenerate_icons proc will now update skin tone and body accessories properly.
rscadd: Organ rejuvenation now clears disfigurement and closes open wounds.
rscadd: Rejuvenating a mob will end all in-progress surgeries.
rscadd: Vox get their cortical stacks back, since the non-implant ones don't interfere with the Heist game-mode.
rscadd: Rejuvenating a long-dead or husked mob will now clear the appropriate mutations.
/:cl:

This might also fix #4404 because of my changes to organ rejuvenation, _but I'm not sure and I haven't checked._

## DO NOT MERGE

I'm testing to see what strange behaviours this might have, if there are any bugs left, and how I can (if I can) condense and clean up the code, etc.

Please leave feedback where you think this can be improved, issues you see, etc. I'll deal with those as usual

I'll post pictures of this in action once I'm back home

Fox-McCloud commented on code/modules/surgery/organs/organ_external.dm:
```diff
@@ -308,14 +308,18 @@ This function completely restores a damaged organ to perfect condition.
 	brute_dam = 0
 	burn_dam = 0
+	open = 0 //Closing all wounds.
+	wounds = list() //Clears all wounds! Good as new.
```
wounds.Cut()

KasparoVy commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1255,16 +1255,55 @@
+/mob/living/carbon/human/proc/check_and_regenerate_organs(var/mob/living/carbon/human/H) //Regenerates missing limbs/organs.
+	var/datum/species/S = H.species
+	var/list/species_organs = S.has_organ
+	var/list/species_limbs = S.has_limbs
```
Going to remove this and the above two lines as they're unnecessary since H.species.has_organ and H.species_has_limbs can be used instead, commit pending

KasparoVy commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1255,16 +1255,55 @@
-	// This will ignore any prosthetics in the prefs currently.
-	species.create_organs(src)
+	//Fix up all organs and replace lost ones.
+	restore_all_organs()
```
Fixes all organs, removes disfigurement, repairs fractures, etc. etc.

KasparoVy commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1255,16 +1255,55 @@
-	species.create_organs(src)
+	//Fix up all organs and replace lost ones.
+	restore_all_organs()
+	check_and_regenerate_organs(src)
```
Replaces missing limbs and organs via comparisons with species-standard limbs+organs and the ones the person getting rejuvenated has.

QuinnAggeler commented:
Will it still remove implants?

tigercat2000 commented:
@KasparoVy could you make sure that however this does it's stuff, it absolutely forces limbs to fix their DNA and colors? because a few nasty bugs are only resolvable right now with rejuv

Fox-McCloud commented:
>Will it still remove implants?

Implants are stored in the mobs contents, not organs.

QuinnAggeler commented:
I mean, the science ones (which are organs codewise). Also, what about implanted objects or bullets/shrapnel?

KasparoVy commented:
@tigercat2000 Mind being a bit more specific? Thanks for commenting, I'll see about sorting that

@QuinnAggeler No, it shouldn't be removing implants (science or otherwise), while shrapnel should be removed by the restore_all_organs() proc (that calls organ rejuvenate() for every organ, internal and external and in turn deletes all implants). I'll try to get a line reference for you in a moment

fludd12 commented:
> Organ rejuvenation now clears disfigurement.
Friend of all IPCs.

KasparoVy commented:
@QuinnAggeler I went ahead and tested it by implanting a Vox with a loyalty implant (from an injector) , an x-ray implant, CNS rebooter and reviver implant in order to cover the different types that you'd mentioned.

![before shot](https://cloud.githubusercontent.com/assets/12377767/15310162/7916d8f6-1bbb-11e6-9460-c4477bf6778d.PNG)

I then shot the Vox a bunch of times, making sure they got an unhealthy dose of bullet shrapnel.

![after shot](https://cloud.githubusercontent.com/assets/12377767/15310198/bbbedfd2-1bbb-11e6-8d17-fdde02a5e2d5.PNG)

I then rejuvenated the Vox and checked in the scanner once again.

![after shot and after rejuvenation](https://cloud.githubusercontent.com/assets/12377767/15310207/cafcbf50-1bbb-11e6-8134-8b1788de2359.PNG)

QuinnAggeler commented:
Cool, all seems in order!

tigercat2000 commented:
> @tigercat2000 Mind being a bit more specific? Thanks for commenting, I'll see about sorting that. As far as I know, when the mob is rejuvenated, hair, head accessory, body colour, body accessory, etc. etc. all that-- those remain untouched. But if the head was removed off, say, an IPC, and then they were rejuvenated, it would give them a default IPC head with the default hair style, head accessory style, etc.

Whenever a mob's skin tone/skin color/eye color needs to be updated, half the update procs work only about half the time. Rejuvenate works every time because it remakes the limbs entirely.

KasparoVy commented:
Huh, alright. I'll change the vars of a bunch of different stuff (ha_style, m_style, h_style, body accessory and colours) and then rejuvenate to see how that fares.

KasparoVy commented:
@tigercat2000 
After making the changes to regenerate_icons and adding the update_eyes() and update_dna() to the rejuvenate proc in life.dm, rejuvenating a player now updates their everything (body colour, eye colour, skin tone, etc) correctly, thanks a bunch for bringing that up

![rejuv icon updates](https://cloud.githubusercontent.com/assets/12377767/15311389/50ced346-1bc8-11e6-8bc0-c4bdea7f0c26.png)

KasparoVy commented:
@tigercat2000 
After making the changes to regenerate_icons and adding the update_eyes() and update_dna() to the rejuvenate proc in life.dm, rejuvenating a player now updates their everything (body colour, eye colour, skin tone, marking style, head accessory style, hair style, facial hair style, etc) correctly, thanks a bunch for bringing that up

Human with the darkest skin tone and white eyes to regular skin tone with blue eyes, and then a totally blacked out Vulpkanin to an eye-searing Vulpkanin
![rejuv icon updates](https://cloud.githubusercontent.com/assets/12377767/15311389/50ced346-1bc8-11e6-8bc0-c4bdea7f0c26.png)

KasparoVy commented:
Currently working at sorting out a logical error whereby mobs with hearts/eyes from other species (i.e. a human with nucleation eyes and heart) will receive an extra set because the eyes and heart are from a different species, but aren't deleted.

KasparoVy commented:
Alright, that should sort it. I tested by putting nucleation eyes into a Vulpkanin and rejuvenating the Vulp before closing up their skull and finishing the surgery.

I checked their vars, and the eyes were in the correct lists. I possessed the mob, and they could see just fine. Their skull wasn't broken and they had no open wounds (which would indicate errors). I switched back to the surgeon and laid the vulpkanin down on the table again and checked the next surgical step to find that rejuvenation correctly ends all surgeries. No problems there.
![surgeryafterrejuv](https://cloud.githubusercontent.com/assets/12377767/15352086/79d1e3e4-1cb0-11e6-8f36-f4bf43ecceb3.PNG)

I even put a debug message in the spots where it changes the 'status' vars of the external and internal organs to make sure they were being set correctly, and that indicated it was working just fine, too.

Next, I did like my commented example and replaced a Vulpkanin's prosthetic arm with a Diona's arm, putting the prosthetic on the Diona afterwards and rejuvenating them both before completing the surgeries with the last steps.
Even with the Diona runtime issue (fixed in another PR), rejuvenation still worked perfectly.
![limbswap](https://cloud.githubusercontent.com/assets/12377767/15352125/b3febfe2-1cb0-11e6-818b-dc1a1336e8a6.PNG)

KasparoVy commented:
Those were the last of the issues I was worried about, but I'll keep rooting around to see if I can find any other quirks/bugs but if I don't, I think this is pretty much done. Still DNM until then

KasparoVy commented:
Something came up. I'll look this over again on Saturday.

Fox-McCloud commented:
@KasparoVy 

Status on this?

KasparoVy commented:
I can't find any more bugs or strange behaviours, but I'd like to request someone else test this. I've never had any experience as an admin and I have no real idea of exactly what kinds of scenarios the rejuvenation verb has been used for.

I've tested it thoroughly, but I want to see how the people who actually use this verb every day find it, in practice, because they might uncover something I missed by using the verb in a way I hadn't, and to get opinions

Kyep commented:
@KasparoVy did you address the two cases I mentioned in Discord?
1) Rejuv doesn't fix prespawned dead clown soldiers that are possessed by an admin and rejuvinated
2) Rejuv doesn't fix hallucinations

KasparoVy commented:
This is the first time I'm hearing of this, thanks for posting it here, I'll commit fixes shortly

KasparoVy commented:
Alright, now husked players and long-dead players that get rejuvenated will get their mutations cleared appropriately.

I tested this on a clown officer. I possessed him, teleported into a safe area and rejuvenated him. He rose from the dead, all damage cleared, and looking like a fresh, normal human.

![clownrises](https://cloud.githubusercontent.com/assets/12377767/15456371/9082d258-203c-11e6-9164-19311e07d00c.PNG)

KasparoVy commented:
You can now normally with objects in your hands again. If you're grabbing someone, you'll flip over them, as intended. No more runtime.

KasparoVy commented:
@Fludd12 If the procedure to re-figure IPC faces is done by screwdrivering, choosing 'cybernetic repair' on the head, screwing, crowbarring, welding and using cable on the head and then crowbarring it shut to close it, re-figuring disfigured IPCs now works again

When I finished the welding stage, the IPC character was no longer disfigured
